### PR TITLE
KVP LHIV TRACKER OFFSITE REGISTER

### PR DESCRIPTION
**Pull request by gabriel090**
Add KVP LHIV tracker offsite register in the EMR to help DICEs track those clients who receive Care and Treatment service from  a different facility other than where they receive KVP services.

The source document for the register is the PLHIV Link Facility form.

### CODE REVIEW

**Review by njorocs — CHANGES_REQUESTED** [review]
Revise the queries to use PLHIV Link Facility Documentation Tracking form as the primary and then treatment verifcation as secondary for accuracy

File: api/src/main/java/org/openmrs/module/kenyakeypop/reporting/cohort/definition/evaluator/KvpLHIVTrackingRegisterCohortDefinitionEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		KvpLHIVTrackingRegisterCohortDefinition definition = (KvpLHIVTrackingRegisterCohortDefinition) cohortDefinition;
+		
+		if (definition == null)
+			return null;
+		
+		Cohort newCohort = new Cohort();
+		
+		context = ObjectUtil.nvl(context, new EvaluationContext());
+		
+		String qry = "select v.client_id from kenyaemr_etl.etl_clinical_visit v\n"
```
[the remark this answers is not in the document]
Comment (gabriel090) [reply api/src/main/java/org/openmrs/module/kenyakeypop/reporting/cohort/definition/evaluator/KvpLHIVTrackingRegisterCohortDefinitionEvaluator.java]: > KPLHIV means KPs living with HIV. This cohort query does not do that.

Worked on It. Thanks

File: api/src/main/java/org/openmrs/module/kenyakeypop/reporting/data/converter/definition/evaluator/kp/FinalTransitionOutcomeDataEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	
+	@Autowired
+	private EvaluationService evaluationService;
+	
+	public EvaluatedPersonData evaluate(PersonDataDefinition definition, EvaluationContext context)
+	        throws EvaluationException {
+		EvaluatedPersonData c = new EvaluatedPersonData(definition, context);
+		
+		String qry = "select fup.encounter_id,\n"
+		        + "       case tr.true_status\n"
```
[the remark this answers is not in the document]
Comment (gabriel090) [reply api/src/main/java/org/openmrs/module/kenyakeypop/reporting/data/converter/definition/evaluator/kp/FinalTransitionOutcomeDataEvaluator.java]: fixed.

File: api/src/main/java/org/openmrs/module/kenyakeypop/reporting/data/converter/definition/evaluator/kp/IPTStartedDateDataEvaluator.java [hunk 3]
```diff
@@ -35,8 +35,15 @@ public class IPTStartedDateDataEvaluator implements PersonDataEvaluator {
 	public EvaluatedPersonData evaluate(PersonDataDefinition definition, EvaluationContext context)
 	        throws EvaluationException {
 		EvaluatedPersonData c = new EvaluatedPersonData(definition, context);
-		String qry = "select v.client_id,mid(max(concat(v.visit_date,v.IPT_start_date)),11) as IPT_start_date  from kenyaemr_etl.etl_treatment_verification v\n"
-		        + "where date(v.visit_date) between date(:startDate) and date(:endDate) group by v.client_id;";
+		String qry = "SELECT  t1.client_id, COALESCE(t1.IPT_start_date, t2.date_initiated_tpt) AS IPT_start_date\n"
```
Comment (njorocs) [inline api/src/main/java/org/openmrs/module/kenyakeypop/reporting/data/converter/definition/evaluator/kp/IPTStartedDateDataEvaluator.java:38]: select from a parent table and left join with these either or tables.

File: api/src/main/java/org/openmrs/module/kenyakeypop/reporting/data/converter/definition/evaluator/kp/OppotunisticInfectiousDiagnosedDateDataEvaluator.java [hunk 4]
```diff
@@ -36,8 +36,15 @@ public EvaluatedPersonData evaluate(PersonDataDefinition definition, EvaluationC
 	        throws EvaluationException {
 		EvaluatedPersonData c = new EvaluatedPersonData(definition, context);
 		
-		String qry = "select v.client_id,mid(max(concat(v.visit_date,v.oi_diagnosis_date)),11) as oi_diagnosis_date  from kenyaemr_etl.etl_treatment_verification v\n"
-		        + "where date(v.visit_date) between date(:startDate) and date(:endDate) group by v.client_id;";
+		String qry = "SELECT  t1.client_id, COALESCE(t1.oi_diagnosis_date, t2.date_diagnosed) AS oi_diagnosis_date\n"
```
Comment (njorocs) [inline api/src/main/java/org/openmrs/module/kenyakeypop/reporting/data/converter/definition/evaluator/kp/OppotunisticInfectiousDiagnosedDateDataEvaluator.java:39]: select from a parent table and left join with these either or tables.

**Review by njorocs — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/kenyakeypop/reporting/cohort/definition/evaluator/KvpLHIVTrackingRegisterCohortDefinitionEvaluator.java [hunk 5]
```diff
@@ -51,10 +51,22 @@ public EvaluatedCohort evaluate(CohortDefinition cohortDefinition, EvaluationCon
 		
 		context = ObjectUtil.nvl(context, new EvaluationContext());
 		
-		String qry = "select v.client_id from kenyaemr_etl.etl_clinical_visit v\n"
-		        + "             inner join kenyaemr_etl.etl_contact c on c.client_id = v.client_id\n"
-		        + "  where c.voided = 0 and date(v.visit_date) BETWEEN DATE(:startDate) AND DATE(:endDate)\n"
-		        + "  group by c.client_id;\n";
+		String qry = "select \n" + "    v.client_id, \n"
```
[the remark this answers is not in the document]
Comment (gabriel090) [reply api/src/main/java/org/openmrs/module/kenyakeypop/reporting/cohort/definition/evaluator/KvpLHIVTrackingRegisterCohortDefinitionEvaluator.java:54]: I have rectified

**Review by njorocs — APPROVED** [review]
merging for testing.

**Pull request merged**